src/link/Elf.zig

```diff
@@ -3104,43 +3107,81 @@ fn initSyntheticSections(self: *Elf) !void {
     const needs_eh_frame = for (self.objects.items) |index| {
         if (self.file(index).?.object.cies.items.len > 0) break true;
     } else false;
-    if (needs_eh_frame) {
-        self.eh_frame_section_index = try self.addSection(.{
-            .name = try self.insertShString(".eh_frame"),
-            .type = elf.SHT_PROGBITS,
-            .flags = elf.SHF_ALLOC,
-            .addralign = ptr_size,
-            .offset = std.math.maxInt(u64),
-        });
-
-        if (comp.link_eh_frame_hdr) {
-            self.eh_frame_hdr_section_index = try self.addSection(.{
-                .name = try self.insertShString(".eh_frame_hdr"),
-                .type = elf.SHT_PROGBITS,
+    if (needs_eh_frame or self.zigObjectPtr() != null) {
+        if (self.eh_frame_section_index == null) {
+            self.eh_frame_section_index = try self.addSection(.{
+                .name = try self.insertShString(".eh_frame"),
+                .type = if (target.cpu.arch == .x86_64)
+                    elf.SHT_X86_64_UNWIND
+                else
+                    elf.SHT_PROGBITS,
                 .flags = elf.SHF_ALLOC,
-                .addralign = 4,
+                .addralign = ptr_size,
                 .offset = std.math.maxInt(u64),
             });
+            if (self.zigObjectPtr()) |zo| {
+                const addSectionSymbol = struct {
+                    fn addSectionSymbol(
+                        zig_object: *ZigObject,
+                        alloc: Allocator,
+                        name: [:0]const u8,
+                        alignment: Atom.Alignment,
+                        shndx: u32,
+                    ) !Symbol.Index {
+                        const name_off = try zig_object.addString(alloc, name);
+                        const index = try zig_object.newSymbolWithAtom(alloc, name_off);
+                        const sym = zig_object.symbol(index);
+                        const esym = &zig_object.symtab.items(.elf_sym)[sym.esym_index];
+                        esym.st_info |= elf.STT_SECTION;
+                        const atom_ptr = zig_object.atom(sym.ref.index).?;
+                        atom_ptr.alignment = alignment;
+                        atom_ptr.output_section_index = shndx;
+                        return index;
+                    }
+                }.addSectionSymbol;
```

Comment:
We already have this logic defined and used in `initMetadata`. Again, we should not call `initSyntheticSections` from `initMetadata`.